File: samples/objc/FirebaseUIChat/FIRSamplesViewController.m
//
//  SamplesViewController.m
//
//  Copyright (c) 2016 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "FIRSample.h"
#import "FIRSamplesViewController.h"
#import "ViewController.h"

@interface FIRSamplesViewController ()

@property (nonatomic) NSArray *samplesContainer;

@end

@implementation FIRSamplesViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  self.navigationController.navigationBar.topItem.title = self.title;
  self.clearsSelectionOnViewWillAppear = NO;

  [self populateSamples];
}

- (void)populateSamples {
  NSMutableArray *samples = [[NSMutableArray alloc] init];

  [samples addObject:[FIRSample sampleWithTitle:@"Auth"
                              sampleDescription:@"Demonstrates the FirebaseAuthUI flow with customization options"
                                     controller:^UIViewController *{
                                       UIViewController *controller =
                                       [[UIStoryboard storyboardWithName:@"Main"
                                                                  bundle:NULL] instantiateViewControllerWithIdentifier:@"FIRAuthViewController"];
                                       return controller;
                                     }]];

  [samples addObject:[FIRSample sampleWithTitle:@"Chat"
                              sampleDescription:@"Demonstrates using a FirebaseCollectionViewDataSource to load data from Firebase Database into a UICollectionView for a basic chat app."
                                     controller:^UIViewController *{
                                       UIViewController *controller =
                                       [[UIStoryboard storyboardWithName:@"Main"
                                                                  bundle:NULL] instantiateViewControllerWithIdentifier:@"ViewController"];
                                       return controller;
                                     }]];


  _samplesContainer = samples;
}

- (void)didReceiveMemoryWarning {
  [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  return _samplesContainer.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  static NSString *cellId = @"FIRSampleCell";
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellId forIndexPath:indexPath];

  FIRSample *sample = _samplesContainer[indexPath.row];
  cell.textLabel.text = sample.title;
  cell.detailTextLabel.text = sample.sampleDescription;


  return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  FIRSample *sample = _samplesContainer[indexPath.row];
  UIViewController *viewController = sample.controllerBlock();
  
  [self.navigationController pushViewController:viewController animated:YES];
}

@end

File: samples/swift/uidemo/Samples/FIRSamplesViewController.swift
//
//  Copyright (c) 2016 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import UIKit

// This controller exists solely to list the samples we've defined thus far.
// Because all of that stuff is static and unchanging, if the app ever crashes
// in here it's probably a typo or some other small accident.
class FIRSamplesViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
  
  private let reuseIdentifier = "FIRSamplesViewControllerCell"
  
  @IBOutlet private var tableView: UITableView!

  override func viewDidLoad() {
    super.viewDidLoad()
    self.tableView.delegate = self
    self.tableView.dataSource = self
    
    // self-sizing cells
    self.tableView.estimatedRowHeight = 85
    self.tableView.rowHeight = UITableViewAutomaticDimension
  }
  
  // MARK: - UITableView Delegate
  
  func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
    let navController = self.navigationController! // assert nonnull
    let targetController = FIRSample(rawValue: indexPath.row)!.controller()
    
    navController.pushViewController(targetController, animated: true)
    tableView.deselectRowAtIndexPath(indexPath, animated: true)
  }
  
  // MARK: - UITableView Data Source

  func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
    let sampleType = FIRSample(rawValue: indexPath.row)!
    
    let cell = self.tableView.dequeueReusableCellWithIdentifier(reuseIdentifier) as! FIRSampleCell
    
    cell.configureWithType(sampleType)
    
    return cell
  }
  
  func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return FIRSample.total
  }
}
